### DISCUSSION TITLE
Is there any risk in registering the ComponentStore at the root level?

### DISCUSSION THREAD

**Discussion by lichunbin814** [ngrx/platform#3498] [Q&A]
I saw [the article](https://zoaibkhan.com/blog/simple-state-management-in-angular-with-ngrx-component-store/), they put the componentStore at the root level as a global store, which is convenient for development and can work normally.

[![enter image description here][1]][1]

However, in the NgRx documentation, [@NgRx/Store](https://ngrx.io/guide/store/why) is still used as an example if you want to manage global state.

[![enter image description here][2]][2]

In my opinion, it seems correct to regard componentStore as a global service, but it conflicts with the design concept of ComponentStore. I don't know if there are any objections. Do you have any opinion?

  [1]: https://i.stack.imgur.com/UicWE.png
  [2]: https://i.stack.imgur.com/5xIEU.png

**Comment by e-oz** [comment]
I’m using it in multiple apps, this approach successfully works even with standalone components.

**Reply by lichunbin814** [reply]
Curious, what are the considerations for choosing componentStore for global state management instead of using store??

**Reply by e-oz** [reply]
Size of the application. Small apps with a few “actions” can avoid a lot of boilerplate by replacing them with a few effects of ComponentStore.

**Reply by Harpush** [reply]
We are using ngrx and considering moving to component store. The main reasons are:
1. When chaining actions across effects the order is not what you would expect. With component store it's depth first.
2. When the app is becoming large you tend to see effects doing two updates due to separation of concerns which result in selectors running twice. In component store you have debounce.
3. Ngrx gives you lots of trouble creating reusable store parts especially with effects. In component store you can use static providers and create fractal composition.
4. Lastly ngrx as local store is not an easy task. Component store can be easily both local and global. And connecting to point 3 the reusable part can be used in both local and global stores now.

**Reply by lichunbin814** [reply]
@Harpush It seems that the store will encounter many problems when changing actions across effects?
 
 But I don't have as much experience as you, can you tell me more practical examples?  
  
Thank you.

**Reply by Harpush** [reply]
The ngrx best practice is each action is unique and dispatched from one place. Many reducer and effects should handle it. The only chain of actions is when you are doing async stuff in effects. In practice it's tough to do it especially when you have many parts which are the same (say grid store for filtering and sorting etc). Once you chain two actions it will probably result in two state updates which will result in two selectors runs and two UI updates.

**Reply by lichunbin814** [reply]
@Harpush Is my understanding the same as what you want to say?

ComponentStore does not have excessive encapsulation, so it still retains the flexibility of Angular DI, allowing us to manage global/local states as if we were using Angular Service.

**Reply by Harpush** [reply]
Indeed - but I wouldn't call it encapsulation... The main difference is the lack of actions as another decouple layer. It has its disadvantages too (at least theoretically) - no dev tools and time travel and all logic is in one place so it might get bloated.

**Reply by lichunbin814** [reply]
@Harpush Regarding the point you mentioned earlier, I still don't understand, the additional benefits of changing to ComponentStore:
 `When the app is becoming large you tend to see effects doing two updates due to separation of concerns which result in selectors running twice. In component store you have debounce.`

--

Because Store's Selector can also use debounce, am I missing any details?
[demo link](https://stackblitz.com/edit/angular-f9hdm6-taxeip?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.component.html)
![image](https://user-images.githubusercontent.com/6021919/180602604-7c0f34d1-0f01-41c3-9aab-3009bf7ea97c.png)

**Reply by lichunbin814** [reply]
@Harpush I think even if it is not CompeontStore, in many Angular Services, there will be a problem of single file becoming bloated.

My current solution: Try to keep the CompeontStore as a single responsibility, and use different CompeontStores to avoid the problem of file bloating.

Do you have any other suggestions?

**Reply by Harpush** [reply]
There are two differences:
1. Component store has custom debounce which doesn't use timers (rxjs asapScheduler).
2. Each selector is debounced not just the end one in the UI which means the whole selectors logic running twice can be prevented. Your example only prevents UI updates.

Maybe there is a way to apply those benefits to ngrx selectors too. Anyway component store selectors are not free of problems either. That's why we wrapped it with a custom wrapper.

**Comment by timdeschryver** [comment]
There's no risk into using component-store as a global store. Projects have different needs, and teams have their own preferences. @ngrx/component-store started as a store for components, but I see more projects using it also as a global store, or in combination with @ngrx/store.
